DICOM series selection abort causes error

When providing a DICOM folder to a command as input, as we all know, the folder(s) get scanned for proper series and you get a "select series" menu, where a series can be selected by typing a number (0-...) and hitting enter. Entering anything else than the valid number results in a 

```
[ERROR] no DICOM series selected
mrinfo: [ERROR] error opening image "FOLDER_YOU_PROVIDED"
```

However, the initial prompt also mentions

```
Select series ('q' to abort):
```

But however then again, entering `q` and hitting enter results in the same above error. As a user, I'd expect no error in this scenario, as it's a valid option.

--- a/cmd/mrinfo.cpp
+++ b/cmd/mrinfo.cpp
@@ -276,40 +276,46 @@ void run ()
       json_keyval || json_all);
 
   for (size_t i = 0; i < argument.size(); ++i) {
-    auto header = Header::open (argument[i]);
-    if (raw_dwgrad)
-      DWI::set_DW_scheme (header, DWI::get_DW_scheme (header));
-    else if (export_grad || check_option_group (GradImportOptions) || dwgrad || shell_bvalues || shell_sizes || shell_indices)
-      DWI::set_DW_scheme (header, DWI::get_valid_DW_scheme (header, true));
+    try {
+      auto header = Header::open (argument[i]);
 
-    if (name)       std::cout << header.name() << "\n";
-    if (format)     std::cout << header.format() << "\n";
-    if (ndim)       std::cout << header.ndim() << "\n";
-    if (size)       print_dimensions (header);
-    if (spacing)    print_spacing (header);
-    if (datatype)   std::cout << (header.datatype().specifier() ? header.datatype().specifier() : "invalid") << "\n";
-    if (strides)    print_strides (header);
-    if (offset)     std::cout << header.intensity_offset() << "\n";
-    if (multiplier) std::cout << header.intensity_scale() << "\n";
-    if (transform)  print_transform (header);
-    if (dwgrad)     std::cout << DWI::get_DW_scheme (header) << "\n";
-    if (shell_bvalues || shell_sizes || shell_indices) print_shells (header, shell_bvalues, shell_sizes, shell_indices);
-    if (petable)    std::cout << PhaseEncoding::get_scheme (header) << "\n";
+      if (raw_dwgrad)
+        DWI::set_DW_scheme (header, DWI::get_DW_scheme (header));
+      else if (export_grad || check_option_group (GradImportOptions) || dwgrad || shell_bvalues || shell_sizes || shell_indices)
+        DWI::set_DW_scheme (header, DWI::get_valid_DW_scheme (header, true));
 
-    for (size_t n = 0; n < properties.size(); ++n)
-      print_properties (header, properties[n][0]);
+      if (name)       std::cout << header.name() << "\n";
+      if (format)     std::cout << header.format() << "\n";
+      if (ndim)       std::cout << header.ndim() << "\n";
+      if (size)       print_dimensions (header);
+      if (spacing)    print_spacing (header);
+      if (datatype)   std::cout << (header.datatype().specifier() ? header.datatype().specifier() : "invalid") << "\n";
+      if (strides)    print_strides (header);
+      if (offset)     std::cout << header.intensity_offset() << "\n";
+      if (multiplier) std::cout << header.intensity_scale() << "\n";
+      if (transform)  print_transform (header);
+      if (dwgrad)     std::cout << DWI::get_DW_scheme (header) << "\n";
+      if (shell_bvalues || shell_sizes || shell_indices) print_shells (header, shell_bvalues, shell_sizes, shell_indices);
+      if (petable)    std::cout << PhaseEncoding::get_scheme (header) << "\n";
 
-    DWI::export_grad_commandline (header);
-    PhaseEncoding::export_commandline (header);
+      for (size_t n = 0; n < properties.size(); ++n)
+        print_properties (header, properties[n][0]);
 
-    if (json_keyval)
-      File::JSON::write (header, *json_keyval, (argument.size() > 1 ? std::string("") : std::string(argument[0])));
+      DWI::export_grad_commandline (header);
+      PhaseEncoding::export_commandline (header);
 
-    if (json_all)
-      header2json (header, *json_all);
+      if (json_keyval)
+        File::JSON::write (header, *json_keyval, (argument.size() > 1 ? std::string("") : std::string(argument[0])));
 
-    if (print_full_header)
-      std::cout << header.description (get_options ("all").size());
+      if (json_all)
+        header2json (header, *json_all);
+
+      if (print_full_header)
+        std::cout << header.description (get_options ("all").size());
+    }
+    catch (CancelException& e) {
+      e.display(-1);
+    }
   }
 
   if (json_keyval) {

--- a/core/exception.h
+++ b/core/exception.h
@@ -117,6 +117,12 @@ namespace MR
         : Exception(previous_exception, msg) {}
   };
 
+
+  class CancelException : public Exception { NOMEMALIGN
+    public:
+      CancelException () : Exception ("operation cancelled by user") { }
+  };
+
   void display_exception_cmdline (const Exception& E, int log_level);
   void cmdline_print_func (const std::string& msg);
   void cmdline_report_to_user_func (const std::string& msg, int type);

--- a/core/file/dicom/select_cmdline.cpp
+++ b/core/file/dicom/select_cmdline.cpp
@@ -105,15 +105,15 @@ namespace MR {
             }
             std::cerr << "? ";
             std::cin >> buf;
-            if (!isdigit (buf[0])) {
-              series.clear();
-              return series;
+            if (buf[0] == 'q' || buf[0] == 'Q')
+              throw CancelException();
+            if (isdigit (buf[0])) {
+              int n = to<int>(buf) - 1;
+              if (n <= (int) tree.size())
+                patient_p = tree[n].get();
             }
-            int n = to<int>(buf) - 1;
-            if (n > (int) tree.size())
+            if (!patient_p)
               fprintf (stderr, "invalid selection - try again\n");
-            else
-              patient_p = tree[n].get();
           }
         }
         else
@@ -145,15 +145,15 @@ namespace MR {
             }
             std::cerr << "? ";
             std::cin >> buf;
-            if (!isdigit (buf[0])) {
-              series.clear();
-              return series;
+            if (buf[0] == 'q' || buf[0] == 'Q')
+              throw CancelException();
+            if (isdigit (buf[0])) {
+              int n = to<int>(buf) - 1;
+              if (n <= (int) patient.size())
+                study_p = patient[n].get();
             }
-            int n = to<int>(buf) - 1;
-            if (n > (int) patient.size())
+            if (!study_p)
               fprintf (stderr, "invalid selection - try again\n");
-            else
-              study_p = patient[n].get();
           }
         }
         else
@@ -188,28 +188,26 @@ namespace MR {
             }
             std::cerr << "? ";
             std::cin >> buf;
-            if (!isdigit (buf[0])) {
-              series.clear();
-              return (series);
-            }
-            vector<int> seq;
-            try {
-              seq = parse_ints (buf);
-            }
-            catch (Exception) {
-              fprintf (stderr, "Invalid number sequence - please try again\n");
-              seq.clear();
-            }
-            if (seq.size()) {
-              for (size_t i = 0; i < seq.size(); i++) {
-                if (seq[i] < 0 || seq[i] >= (int) study.size()) {
-                  fprintf (stderr, "invalid selection - try again\n");
-                  series.clear();
-                  break;
+            if (buf[0] == 'q' || buf[0] == 'Q')
+              throw CancelException();
+            if (isdigit (buf[0])) {
+              vector<int> seq;
+              try {
+                seq = parse_ints (buf);
+                for (size_t i = 0; i < seq.size(); i++) {
+                  if (seq[i] < 0 || seq[i] >= (int) study.size()) {
+                    series.clear();
+                    break;
+                  }
+                  series.push_back (study[seq[i]]);
                 }
-                series.push_back (study[seq[i]]);
+              }
+              catch (Exception) {
+                seq.clear();
               }
             }
+            if (series.size() == 0)
+              fprintf (stderr, "Invalid selection - please try again\n");
           }
         }
         else series.push_back (study[0]);

--- a/core/header.cpp
+++ b/core/header.cpp
@@ -271,6 +271,7 @@ namespace MR
       if (do_realign_transform)
         H.realign_transform();
     }
+    catch (CancelException& e) { throw; }
     catch (Exception& E) {
       throw Exception (E, "error opening image \"" + image_name + "\"");
     }

--- a/src/gui/dialog/dicom.cpp
+++ b/src/gui/dialog/dicom.cpp
@@ -221,6 +221,8 @@ namespace MR
             }
           }
         }
+        else
+          throw CancelException();
         return ret;
       }
 

--- a/src/gui/mrview/window.cpp
+++ b/src/gui/mrview/window.cpp
@@ -774,6 +774,10 @@ namespace MR
           vector<std::unique_ptr<MR::Header>> list;
           for (size_t n = 0; n < MR::App::argument.size(); ++n) {
             try { list.push_back (make_unique<MR::Header> (MR::Header::open (MR::App::argument[n]))); }
+            catch (CancelException& e) {
+              for (const auto& msg : e.description)
+                CONSOLE (msg);
+            }
             catch (Exception& e) { e.display(); }
           }
           add_images (list);
@@ -857,6 +861,9 @@ namespace MR
           vector<std::unique_ptr<MR::Header>> list;
           list.push_back (make_unique<MR::Header> (MR::Header::open (folder)));
           add_images (list);
+        }
+        catch (CancelException& E) {
+          E.display (-1);
         }
         catch (Exception& E) {
           E.display();

